Rework lasagne.objectives (remove Objective and MaskedObjective)

f0k opened the pull request:
This reworks the `lasagne.objectives` module as per our discussion in #173.
Todo list:
* [x] refactor, then deprecate Objective classes
* [x] add module docstring
* [x] add tests
* ~~update occurrences of Objective in the documentation~~ (wasn't mentioned anywhere else)
* [x] update MNIST example

f0k commented:
Ready for reviewing and merging! (Any code using `Objective` or `MaskedObjective` works as before, but remember to notify mailing list when merged.)

benanne commented:
Nice, I'll try to review it in detail this weekend.

Any reason you didn't include `make_var_like()` in one go? Seems like it would only add a few lines? But fair enough, we can separate it if you prefer it that way. Hopefully we won't forget about it though.

f0k commented:
> Any reason you didn't include `make_var_like()` in one go? Seems like it would only add a few lines?

The current example advocates just using `theano.tensor.matrix()` right away, and I don't necessarily think we should promote using `make_var_like()` -- it only works when the targets are the same shape as the outputs. So I'd just have added `make_var_like()` without changing the example, but then it's fully independent of what this PR does.

I'll be abroad from tomorrow until June 15, so please feel free to merge when you reviewed it! The only thing I'm unsure about is whether we want `weighted_aggregate` to be a separate function. We could just call that one `aggregate` and remove the simpler one. The `weights` parameter is optional anyway.

benanne commented on examples/mnist.py:
```diff
@@ -123,16 +123,17 @@ def create_iter_functions(dataset, output_layer,
-        loss_function=lasagne.objectives.categorical_crossentropy)
+    output = lasagne.layers.get_output(output_layer, X_batch)
+    loss_train = lasagne.objectives.categorical_crossentropy(output, y_batch)
+    loss_train = lasagne.objectives.aggregate(loss_train, 'mean')
```
I don't really see the value here of using `aggregate` instead of just writing `loss_train = T.mean(loss_train)`. If I understand correctly it does exactly the same thing anyway. Sure, `aggregate` is more flexible, but since we just want the mean here, why not write that directly so people realise everything is just passing around Theano expressions that they can do with as they please.

benanne commented:
I think getting rid of `aggregate` and renaming `weighted_aggregate` to `aggregate` is a good idea. As it stands, this function does nothing but wrap either `T.mean` or `T.sum`, and needlessly wrapping Theano is something we said we wouldn't do :)

One situation where it is useful is when you want to switch between different aggregation modes, then it's just a question of changing the `mode` argument, which is nice. But for that argument to make sense, `weighted_aggregate` should be the main function anyway.

That's really all I can think of, the docstrings and examples are great and the tests look good as well! Not sure what to do now since you'll be away for a week... maybe I'll amend it myself before merging. Probably not worth waiting till you're back.

f0k commented:
> Not sure what to do now since you'll be away for a week... maybe I'll amend it myself before merging. Probably not worth waiting till you're back.

What's the state on this? Shall I rebase, remove `aggregate` and rename `weighted_aggregate` to `aggregate`?

benanne commented:
Go for it, I haven't done anything.

f0k commented:
Rebased, changed MNIST example to just call `mean()` directly instead of using `aggregate()`, and merged `weighted_aggregate()` into `aggregate()`. Ready to merge when Travis passed. This should also raise coverage to 100%.

f0k commented:
Travis passed. @benanne, please merge when you're fine with it!

f0k commented:
Ah wait, missed an occurrence of `weighted_aggregate`...

f0k commented:
Ready to merge again, this time for real.

benanne commented:
Yay!

benanne commented:
Are you doing the [interface-change] post or should I do it? :)

f0k commented:
Will do it.
